tests/utils/exec.ts

```diff
@@ -0,0 +1,600 @@
+import type { ChildProcess } from 'child_process';
+import type ErrorPolykey from '@/ErrorPolykey';
+import childProcess from 'child_process';
+import fs from 'fs';
+import path from 'path';
+import process from 'process';
+import readline from 'readline';
+import * as mockProcess from 'jest-mock-process';
+import mockedEnv from 'mocked-env';
+import nexpect from 'nexpect';
+import Logger from '@matrixai/logger';
+import main from '@/bin/polykey';
+
+type ExecOpts = {
+  env: Record<string, string | undefined>;
+  command?: string | undefined;
+  cwd?: string;
+  shell?: boolean;
+};
+
+const tsConfigPath = path.resolve(
+  path.join(globalThis.projectDir ?? '', 'tsconfig.json'),
+);
+
+const polykeyPath = path.resolve(
+  path.join(globalThis.projectDir ?? '', 'src/bin/polykey.ts'),
+);
```

Comment:
`globalThis.projectDir ?? ''` had to be added to prevent an error for `globalThis.projectDir` being undefined. Likely the same needs to be done for `globalThis.testDir` somewhere.

Reply:
Why would this be affected by https://github.com/MatrixAI/Polykey/pull/411#issuecomment-1212635793?

This problem was due to `ts-node` running an internal test script that did not have the test environment.

The docker tests I feel like has too much required environment variables, so we want to remove them as much as  possible. At least the `PK_TEST_PLATFORM` should disappear with #435.

Reply:
The problem was that `globalThis` variables are not defined when running as a sub process. That means it won't be defined when running as docker as well. If the docker image is trying to use `globalThis.testDir` somewhere then we will run into an error `globalThis.testDir` is undefined. As far as the type system knows  the `globalThis` variables are always defined so it's not caught when building.

Reply:
What docker process would be using `globalThis`? It would only be possible if you were running a test script under `tests` dir that imports the `tests/utils`. The docker tests are not running test scripts are they? They should be running `src/bin/polykey.ts` which wouldn't be using `globalThis.testDir`...

Reply:
It shouldn't be using `globalThis`. But if this problem is anything like the previous problem where just importing `tests/utils/exec.ts` was causing `global.this` variables to be used then we would get this error. I'll have to dig into why this is happening but it's too similar to the previous problem to not check for that first.

Reply:
Let's do a quickfix for this. But long term we need to move all the integration tests into a separate directory and discard these hacks. Instead we would do things like:

```
import parameterisedTest from '..somecommontestmodule..';

describe('', () => {
  test('', parameterisedTest(p1, p2));
});
```

The idea is that `parameterisedTest(p1, p2)` should return an `() => Promise<void>` function to execute as the actual test.

This way we separate the "test structure" from the actual test itself which can be shared among different test files.

Reply:
I've narrowed it down. the `--node-path` parameter we're giving in the test is not properly relative to the working directory. Leading us to try and create the agent at a path that doesn't exist. 

```
// docker
// node path
'tmp/test/polykey-test-NCPSFm/polykey'

// working directory
/home/faulkes/matrixCode/polykey/js-polykey/tmp/test/polykey-test-NCPSFm
```

```
// normally
// node path
'/tmp/polykey-test-WWjNME/polykey'

// working directory
/tmp/polykey-test-WWjNME
```

Ah I see, normally we're using a full absolute path, but with the docker testing we're using a relative path. To fix this we should be using consistent in the style of path for both cases. The simplest change would be to make dataDir an absolute path in all cases by using `path.resolve`.

Also the `testCmd` was not getting passed to the `pkSpawn` opts.

Reply:
There's a single test failure now. 

```
FAIL tests/bin/polykey.test.ts
  ● polykey › format option affects STDERR
    SyntaxError: Unexpected token d in JSON at position 0
        at JSON.parse (<anonymous>)
      63 |       rlErr.once('close', reject);
      64 |     });
    > 65 |     const stderrParsed = JSON.parse(stderrStart);
         |                               ^
      66 |     expect(stderrParsed).toMatchObject({
      67 |       level: expect.stringMatching(/INFO|WARN|ERROR|DEBUG/),
      68 |       key: expect.any(String),
      at Object.<anonymous> (tests/bin/polykey.test.ts:65:31)
      
```

This is odd, I'm not sure what's causing it and it's passing locally. I'm running the test again with logging enabled to try and sus it out. But if I can't see a quick solution I'll just disable the test.

Reply:
Found and fixed the problem. Resolving this.